### PR TITLE
Fix Use Span Based String Concat violations in System.Diagnostics.DiagnosticSource

### PR DESCRIPTION
**Pull request by NewellClark**
While working on [this analyzer PR](https://github.com/dotnet/roslyn-analyzers/pull/4764) for [#issue 33777](https://github.com/dotnet/runtime/issues/33777) I found 3 violations in `System.Diagnostics.DiagnosticSource`.

### CODE REVIEW

**Comment by NewellClark** [comment]
Area label should be `System.Diagnostics.DiagnosticSource`

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 1]
```diff
@@ -1213,7 +1214,11 @@ private string AppendSuffix(string parentId, string suffix, char delimiter)
 
             //generate overflow suffix
             string overflowSuffix = ((int)GetRandomNumber()).ToString("x8");
+#if SPAN_BASED_STRING_CONCAT
+            return string.Concat(parentId.AsSpan(0, trimPosition), overflowSuffix, stackalloc char[] { '#' });
```
Comment (tarekgh) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1218]: why you are using stackalloc?

Comment (NewellClark) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1218]: There's no implicit conversion from `char` to `ReadOnlySpan<char>`. The alternative is to convert it to a string.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1218]: > but I guess it makes more sense just to use a string literal.

I agree.

**Comment by tarekgh** [comment]
@NewellClark usually I am not a fan of introducing a new def just to call different APIs. did we think either turning off the fired warning? or just implement the concatenation using Span APIs that is supported in non .NET Core platforms too?

**Comment by tarekgh** [comment]
CC @noahfalk

**Comment by NewellClark** [comment]
> @NewellClark usually I am not a fan of introducing a new def just to call different APIs. did we think either turning off the fired warning? or just implement the concatenation using Span APIs that is supported in non .NET Core platforms too?

@tarekgh  Fair enough. The code I changed needs to compile on versions as early as netstandard1.3, which doesn't have the span-based `string.Concat` method. I suppose I could use unsafe code to modify the string after allocating it, to avoid extra copies.

**Comment by tarekgh** [comment]
> Edit: actually I think it may be best to disable the warning here. The warning is for new APIs where span-based string.Concat is available. Since it's not available here, probably simplest just to disable the warning, rather than cluttering things up with unsafe code.

agree.

**Comment by stephentoub** [comment]
> I suppose I could use unsafe code to modify the string after allocating it

Please don't. :wink:

I suggest we just not change anything here.  None of these look like they're on a hot enough path to be worth the hassle.

**Comment by stephentoub** [comment]
> actually I think it may be best to disable the warning here

There shouldn't be anything to disable.  The analyzer should only be firing at an Info level by default; it won't show up as a warning or error in the build once the analyzer is consumed into dotnet/runtime, just as a helpful suggestion for someone using Visual Studio.

**Comment by NewellClark** [comment]
@stephentoub good to know. I was under the impression that it was required to fix or suppress all rules due to [this](https://github.com/dotnet/roslyn-analyzers/blob/master/docs/NetCore_GettingStarted.md#definition-of-done), but if suppression is unnecessary, then feel free to close this PR.

**Comment by stephentoub** [comment]
Thanks, @NewellClark.  @jeffhandley may be interested in fixing those instructions based on your feedback.

**Pull request closed without merging**